lib/controllers/agendash.js

```diff
@@ -56,69 +57,117 @@ module.exports = function (agenda, options) {
-              ],
-            },
-            failed: {
-              $and: [
-                "$lastFinishedAt",
-                "$failedAt",
-                { $eq: ["$lastFinishedAt", "$failedAt"] },
-              ],
-            },
-            repeating: {
-              $and: ["$repeatInterval", { $ne: ["$repeatInterval", null] }],
-            },
+      },
+      {
+        $project: {
+          job: "$$ROOT",
+          _id: "$$ROOT._id",
+          running: {
+            $and: ["$lastRunAt", { $gt: ["$lastRunAt", "$lastFinishedAt"] }],
+          },
+          scheduled: {
+            $and: ["$nextRunAt", { $gte: ["$nextRunAt", new Date()] }],
+          },
+          queued: {
+            $and: [
+              "$nextRunAt",
+              { $gte: [new Date(), "$nextRunAt"] },
+              { $gte: ["$nextRunAt", "$lastFinishedAt"] },
+            ],
+          },
+          completed: {
+            $and: [
+              "$lastFinishedAt",
+              { $gt: ["$lastFinishedAt", "$failedAt"] },
+            ],
+          },
+          failed: {
+            $and: [
+              "$lastFinishedAt",
+              "$failedAt",
+              { $eq: ["$lastFinishedAt", "$failedAt"] },
+            ],
+          },
+          repeating: {
+            $and: ["$repeatInterval", { $ne: ["$repeatInterval", null] }],
           },
         },
-        { $match: postMatch },
-        {
-          $facet: {
-            pages: [
-              { $count: "totalMatchs" },
-              {
-                $project: {
-                  totalPages: {
-                    $ceil: { $divide: ["$totalMatchs", options.limit] },
-                  },
+      },
+      { $match: postMatch }
+
+      
+    ];
+    const standardMongoDBQuery = [
+
+      {
+        $facet: {
+          pages: [
+            { $count: "totalMatchs" },
+            {
+              $project: {
+                totalPages: {
+                  $ceil: { $divide: ["$totalMatchs", options.limit] },
                 },
               },
-            ],
-            filtered: [{ $skip: options.skip }, { $limit: options.limit }],
+            },
+          ],
+          filtered: [{ $skip: options.skip }, { $limit: options.limit }],
+        }
+      }
+    ]
+    const awsDocumentDBQuery = [
+      {
+        $group: {
+          _id: null,
+          totalMatchs: {
+            $sum: 1
           },
-        },
-      ])
-      .toArray();
+          results: {
+            $push: '$$ROOT'
+          }
+        }
+      },
+      {
+        $project: {
+          pages: [{totalPages: '$totalMatchs'}],
+          filtered: {
+            $slice: [
+              '$results',
+              options.skip,
+              options.limit
+            ]
+          }
+        }
+      }
+    ]
+    return collection
+    .aggregate([...aggregateQuery,...standardMongoDBQuery])
+    .toArray()
+    .then((result) => {
+      return result;
+    })
+    .catch((error) => { 
```

Comment:
This is gross, it's an extra round trip to AWS every time we run this query, and I'm not sure amazon will thank us for that. Can we just make this a configuration option (or somehow smartly detect the version of Mongo, rather than this awful thing?)

We're not **even** looking at the error response to detect why it failed :face_exhaling:

Reply:
I'm hoping to avoid us maintaining our own code until upstream implements this properly. This works and it's only doing this query every so and so seconds. I hardly think AWS will notice.

We can reimplement this patch but I'm going to postpone it after EMS staging is migrated.

Reply:
I'm hoping to avoid us maintaining our own code until upstream implements this properly. This works and it's only doing this query every so and so seconds. I hardly think AWS will notice.

We can reimplement this patch but I'm going to postpone it after staging migration, if ok.

Reply:
To clarify the reason for not spending engineering time now: when upstream implements documentdb support, we're going to want to use that implementation anyway, not whatever we came up with. As such I'd avoid making anything fancy at this stage. Given we only need documentdb support, I could simplify the pr - but that carries the issue if the upstream patch gets new commits, we'll be in a merge conflict against that.